exporter/trace/jaeger/jaeger.go

```diff
@@ -0,0 +1,340 @@
+func StringTag(key string, value string) Tag {
+	return Tag{key, value}
+}
+
+// Int64Tag creates a new tag of type int64, exported as jaeger.TagType_LONG
+func Int64Tag(key string, value int64) Tag {
+	return Tag{key, value}
+}
+
+// Exporter is an implementation of trace.Exporter that uploads spans to Jaeger.
+type Exporter struct {
+	endpoint      string
+	agentEndpoint string
+	process       *gen.Process
+	bundler       *bundler.Bundler
+	client        *agentClientUDP
+
+	username, password string
+}
+
+var _ trace.Exporter = (*Exporter)(nil)
+
+// ExportSpan exports a SpanData to Jaeger.
+func (e *Exporter) ExportSpan(data *trace.SpanData) {
+	_ = e.bundler.Add(spanDataToThrift(data), 1)
+	// TODO(jbd): Handle oversized bundlers.
+}
+
+func spanDataToThrift(data *trace.SpanData) *gen.Span {
+	tags := make([]*gen.Tag, 0, len(data.Attributes))
+	for k, v := range data.Attributes {
+		tag := attributeToTag(k, v)
+		if tag != nil {
+			tags = append(tags, tag)
+		}
+	}
+
+	tags = append(tags,
+		attributeToTag("status.code", int32(data.Status)),
+		attributeToTag("status.message", data.Status.String()),
+	)
+
+	// Ensure that if Status.Code is not OK, that we set the "error" tag on the Jaeger span.
+	// See Issue https://github.com/census-instrumentation/opencensus-go/issues/1041
+	if data.Status != codes.OK {
+		tags = append(tags, attributeToTag("error", true))
+	}
+
+	var logs []*gen.Log
+	for _, a := range data.MessageEvents {
+		fields := make([]*gen.Tag, 0, len(a.Attributes()))
+		for _, kv := range a.Attributes() {
+			tag := attributeToTag(kv.Key.Variable.Name, kv.Value.Emit())
+			if tag != nil {
+				fields = append(fields, tag)
+			}
+		}
+		fields = append(fields, attributeToTag("message", a.Message()))
+		logs = append(logs, &gen.Log{
+			//Timestamp: a.Time.UnixNano() / 1000,
+			//TODO: [rghetia] update when time is supported in the event.
+			Timestamp: time.Now().UnixNano() / 1000,
+			Fields:    fields,
+		})
+	}
+	//TODO: [rghetia] add links.
+	//
+	//var refs []*gen.SpanRef
+	//for _, link := range data.Links {
+	//	refs = append(refs, &gen.SpanRef{
+	//		TraceIdHigh: bytesToInt64(link.TraceID[0:8]),
+	//		TraceIdLow:  bytesToInt64(link.TraceID[8:16]),
+	//		SpanId:      bytesToInt64(link.SpanID[:]),
+	//	})
+	//}
+
+	return &gen.Span{
+		TraceIdHigh:   int64(data.SpanContext.TraceID.High),
+		TraceIdLow:    int64(data.SpanContext.TraceID.Low),
+		SpanId:        int64(data.SpanContext.SpanID),
+		ParentSpanId:  int64(data.ParentSpanID),
+		OperationName: data.Name, // TODO: if span kind is added then add prefix "Sent"/"Recv"
+		Flags:         int32(data.SpanContext.TraceOptions),
+		StartTime:     data.StartTime.UnixNano() / 1000,
+		Duration:      data.EndTime.Sub(data.StartTime).Nanoseconds() / 1000,
+		Tags:          tags,
+		Logs:          logs,
+		// TODO: goes with Links.
+		// References:    refs,
+	}
+}
+
+func attributeToTag(key string, a interface{}) *gen.Tag {
```

Comment:
Seems like we could avoid the `interface{}` conversation here by passing through the underlying value, i.e., not Emitting it yet. This would avoid some allocation. Fine to leave a TODO.

Reply:
added a TODO for now.